feat: import event by url

egemengunel opened the pull request:
This PR adds the functionality for event importing from Facebook. The implementation uses Trigger.dev environment for scraping and saving the data to the database.

## Setup Instructions

### Environment Configuration
1. Get your trigger secret key and place it in the `.env` file
2. Go to project settings, copy the `project_ref`, and replace the project value in the `trigger.config.ts` file
3. Add the following environment variables to the `.env` file in `/cli` directory:
   - `NUXT_CLOUDINARY_API_KEY`
   - `NUXT_CLOUDINARY_API_SECRET`
   - `NUXT_PUBLIC_CLOUDINARY_CLOUD_NAME`
   - `GOOGLE_MAPS_API_KEY`

### Running the Trigger
To use the trigger, run `pnpm dlx trigger.dev@latest dev` on a separate terminal. This allows you to run the import functionality locally.

## CLI Tools

CLI tools were added to see the outputs after scraping and saving to the database via the terminal.

### Event Import Debug
To execute `event:import:debug <url>`:
```bash
pnpm cli-tsx event:import:debug <url>
```

**Example:**
```bash
pnpm cli-tsx event:import:debug https://www.facebook.com/events/1129152422074492/1129152468741154/
```

### Event Import
To run `event:import <id>`:

1. Add an event to the database (can be done from the website)
   - Admin area > Events > Create Event > Create New Event
2. Add `sourceUrl` to the event you created using Prisma Studio
3. Run:
   ```bash
   pnpm cli-tsx event:import <event-id-goes-here>
   ```

Closes #118

egemengunel committed:
fix Issue 402 Events Load More

Implement fetchUntilEvents function to improve event loading logic and update load more button behavior in EventsList component

egemengunel committed:
revert EventsList and  add next event fetching on events router

Update the events router to add a shared `commonWhere` filter. Same rules (city, style, type,
  search) in one place, spread into both queries

  add next query that actually looks for the earliest future event rather than the next day, stopping the list from flickering

egemengunel committed:
refactor: remove unused code, remove comments

egemengunel committed:
Create Initial Implementation of Trigger

Add Trigger package and set up a trigger project that auto capitalizes the events that are created

egemengunel committed:
feat: add frontend for Import Page

Integrate EventImportDialog into the user dialog component and modify the event handling logic in the admin events page to use the new import dialog instead of the create event dialog. Update button accordingly.

egemengunel committed:
feat: add event import route to events.ts add form validation to Import.vue

Add form validation and submission logic for importing events from a URL in the Import.vue component. Update the events router to handle the import mutation

razbakov commented on server/trpc/routers/events.ts:
```diff
@@ -319,4 +307,33 @@ export const eventsRouter = router({
+    .input(
+      z.object({
+        sourceUrl: z.string().url(),
+        startDate: z.string().optional(),
```
importing event needs only sourceUrl, please remove startDate and endDate

razbakov commented on server/trpc/routers/events.ts:
```diff
@@ -319,4 +307,33 @@ export const eventsRouter = router({
+          startDate: startDate ? new Date(startDate) : undefined,
+          endDate: endDate ? new Date(endDate) : undefined,
+          creatorId: ctx.session?.profile?.id,
+          organizerId: ctx.session?.profile?.id,
```
organizer should be imported from the url, the person who is adding event is not necessary to be it's organizer

razbakov commented on server/trpc/routers/events.ts:
```diff
@@ -319,4 +307,33 @@ export const eventsRouter = router({
+      })
+      await tasks.trigger('event-name-will-be-here', {
+        eventId: event.id,
+        sourceUrl,
```
i would store sourceUrl in the database, and pass only eventId to the task

egemengunel committed:
feat(admin/events): convert import dialog into full page & store sourceUrl

moved Import.vue to pages/admin/events/import.vue; old dialog removed, button now routes to /admin/events/import.

events.import mutation now accepts only sourceUrl; dropped startDate, endDate, organizerId.

added sourceUrl String? (and made organizerId optional) to Event model; new Prisma migration generated.

prisma.event.create now saves sourceUrl, leaves organizer/date null.

removed EventImportDialog registration and all useDialog calls.

egemengunel committed:
Move EventCreateDialog to new.vue

egemengunel committed:
Import Linguist.ts  from v3

egemengunel committed:
import: add Google Maps integration and Cloudinary utility functions

- imported Google Maps utility functions for place search and city ID retrieval using the @googlemaps/google-maps-services-js package.
- imported Cloudinary utility function to handle image uploads.
- Updated package.json and pnpm-lock.yaml to include new dependencies
- Refactored event trigger name in events router for clarity.

egemengunel committed:
import: facebook_import.ts, rename googleMaps.ts to google_maps.ts

egemengunel committed:
import: import schema_imports.ts  file for handling non facebool events

egemengunel committed:
add: add scraping task

- Updated the event creation logic to include the slug in the database.
- added the scraping task to handle both Facebook and schema events.

egemengunel committed:
cleanup

egemengunel committed:
add styles

egemengunel committed:
import dance_styles, style_utils

egemengunel committed:
import getSuggestedStyles() function in linguist.ts

egemengunel committed:
Import suggested styles

- Integrated getSuggestedStyles function.
- Streamlined the getOrg function for better profile retrieval and creation.

egemengunel committed:
remove static .yml danceStyles files

egemengunel committed:
Remove dance_styles.ts file and refactor style_utils to utilize database for dance styles retrieval. Updated getSuggestedStyles to be asynchronous and adjusted imports accordingly.

razbakov commented on pages/admin/events/index.vue:
```diff
@@ -103,7 +98,7 @@ const columns: ColumnDef<any>[] = [
     <div class="flex items-center gap-2">
       <Input v-model="searchQuery" placeholder="Search events..." />
-      <Button @click="handleCreateEvent"> Create Event </Button>
+      <Button @click="handleImportEvent"> Create Event </Button>
```
Please use NuxtLink

razbakov commented on server/utils/cloudinary.ts:
```diff
@@ -0,0 +1,17 @@
+  }
+
+  cloudinary.config({
+    cloud_name: process.env.CLOUDINARY_CLOUD_NAME as string,
```
please use `useRuntimeConfig()`, see server/api/webhook/mailgun.ts for example

razbakov commented on server/utils/facebook_imports.ts:
```diff
@@ -0,0 +1,211 @@
+      return {
+        id: existingProfile.id,
+        username: existingProfile.username,
+        name: existingProfile.name || existingProfile.username || '',
```
no need to return all this information, just an id is enough

egemengunel committed:
rename facebook_imports, schema_imports to facebook_import, schema_import

razbakov commented on server/utils/style_utils.ts:
```diff
@@ -0,0 +1,6 @@
+import { prisma } from '~/server/prisma'
+
```
seems there is no need for this file, you can just add that code where it's used. it's just one line.

razbakov commented on server/utils/linguist.ts:
```diff
@@ -0,0 +1,128 @@
+  const styles: any = {}
+
+  for (const style of stylesList) {
+    if (!style.regexp) {
```
that can be filtered by prisma

razbakov replied:
if already filtered in prisma, no need to keep this condition

razbakov commented on server/trpc/routers/events.ts:
```diff
@@ -319,4 +306,32 @@ export const eventsRouter = router({
+          slug: placeholderSlug,
+        },
+      })
+      await tasks.trigger('scrape-and-fill-events', {
```
rename to `import-event`

razbakov commented on pages/admin/events/new.vue:
```diff
@@ -0,0 +1,59 @@
+
+<template>
+  <div class="min-h-screen flex items-center justify-center ml-8 mr-8">
+    <form @submit="onSubmit" class="w-[650px]">
```
pages shouldn't contain complex logic, logic should be inside components

razbakov commented on pages/admin/events/new.vue:
```diff
@@ -0,0 +1,59 @@
+const router = useRouter()
+
+const props = defineProps<{
+  onSuccess?: () => void
```
pages don't have props, also this is not used.

egemengunel committed:
- Replaced the button for creating events with NuxtLink
- Removed unused props from the new event component.
- Updated the event import task ID.

egemengunel committed:
Refactor cloudinary config to use runtime config and clean up getOrg function by removing unused profile properties.

egemengunel committed:
Refactor Cloudinary integration to use dependency injection for configuration. Updated image upload functions to accept Cloudinary config as a parameter. Removed style_util and implemented the function inside linguist.ts

egemengunel committed:
Update getSuggestedStyles to retrieve dance styles from the database using Prisma. Simplified word extraction from regex by filtering out empty values.

egemengunel committed:
create NewEventForm.vue component

egemengunel committed:
Commponentize Import Page

egemengunel committed:
Switch To NuxtLink in NewEventForm

egemengunel committed:
fix routing in NewEventForm

egemengunel committed:
Add success toast notification in ImportForm and update error message in NewEventForm

egemengunel committed:
remove GetStyles func

since now prisma is handling  getSuggestedStyles, we dont need getStyles()

razbakov commented on components/event/ImportForm.vue:
```diff
@@ -0,0 +1,69 @@
+    })
+    toast.success('Your import has started.')
+  } catch (e: any) {
+    console.log(e.message ?? 'Import Failed')
```
error messages should be displayed to a user using toast

razbakov commented on components/event/ImportForm.vue:
```diff
@@ -0,0 +1,69 @@
+    </div>
+
+    <NuxtLink to="/admin/events/new">
+      <Button variant="outline" class="w-full mt-4">Create new event</Button>
```
You should use
```
<Button as-child>
<NuxtLink>
```

See last section in https://www.shadcn-vue.com/docs/components/button.html

razbakov commented on nuxt.config.ts:
```diff
@@ -112,8 +112,8 @@ export default defineNuxtConfig({
     stripeWebhookSecret: '',
-    cloudinaryApiKey: '',
-    cloudinaryApiSecret: '',
+    cloudinaryApiKey: process.env.CLOUDINARY_API_KEY,
```
as you see in other fields, no need to refer to `process.env`, just leave it empty. more about it here: https://nuxt.com/docs/guide/going-further/runtime-config also see attachment "Watch a video from Alexander Lichter showcasing the top mistake developers make using runtimeConfig."

razbakov commented on pages/admin/events/import.vue:
```diff
@@ -0,0 +1,5 @@
+<template>
+  <div class="min-h-screen flex items-center justify-center ml-8 mr-8">
+    <ImportForm />
```
Import what? `<ImportEventForm>`, and then similar to that the other form would be `<AddEventForm>`

razbakov commented on server/utils/cloudinary.ts:
```diff
@@ -0,0 +1,25 @@
+  }
+
+  cloudinary.config({
+    cloud_name: config.cloud_name,
```
This config should come from runtime config

razbakov commented on server/utils/google_maps.ts:
```diff
@@ -0,0 +1,88 @@
+    place = await client.placeDetails({
+      params: {
+        place_id: String(response.data.results[0].place_id),
+        key: String(process.env.GOOGLE_MAPS_API_KEY),
```
should be from runtime config

egemengunel committed:
Add error toast notification in ImportForm and rename AddEventForm and ImportEventForm components

egemengunel committed:
Refactor button  in ImportEventForm to use NuxtLink as a child of Button.

egemengunel committed:
Refactor slug  by replacing slugify with getSlug in event creation and Facebook import functions.

egemengunel committed:
Move: move cloudinary.ts, facebook_import.ts, linguist.ts, schema_import.ts, from server/utils to trigger/scraper/ functions,
Move: scrape.ts from trigger to trigger/scraper

egemengunel committed:
revert cloudinary.ts use process.env

egemengunel committed:
refactor linguist.ts to remove unnecesary regexp value check, update env.example

egemengunel committed:
Refactor buttons to use NuxtLink as a child of Button for improved structure and consistency.

egemengunel committed:
Refactor cloudinary.ts to ensure environment variables use String()

egemengunel committed:
Componentize scrape.ts by seperating getEventType and saveEvent and getRedirectedUrl functions

egemengunel committed:
add deletePlaceHolderEvent func in case of import failure to delete the unused row, rename save_event to manage_event

egemengunel committed:
sync database with the schema add importError field to events model

egemengunel committed:
Refactor scrape and manage_event functions to replace deletePlaceholderEvent with handleImportFailure to store error messages in the database

egemengunel committed:
Fix typo in import error message and update logic in manage_event.ts to set default values on import failure.

egemengunel committed:
Refactor error messages

egemengunel committed:
revert prisma.schema

egemengunel committed:
add importError  field to event model

egemengunel committed:
uodate env to include TRIGGER_PROJECT_REF, import and new pages now use the admin page layout.

egemengunel committed:
Update import success and error messages, remove error logging from getRedrictedUrl function

egemengunel committed:
revert trigger config project refrence to be hardcoded

egemengunel committed:
add script setup, remove import on new.vue

egemengunel committed:
Update env variables for Cloudinary, update trigger configuration with Prisma extension

egemengunel committed:
Add Puppeteer extension to trigger configuration and add creatorID to the list of events to fetch in the admin panel

egemengunel committed:
Add optional 'type' field to event model in events router

razbakov committed:
update trigger project key

razbakov committed:
admin: limit to 10 events

razbakov committed:
admin: sort events by created at

razbakov committed:
admin: event forms with toast promises

razbakov committed:
event import: fix event type

razbakov committed:
import event: ticket url

razbakov committed:
event: slug not required

egemengunel committed:
fix: decode Facebook username in getOrg function

egemengunel committed:
refactor: make facebook_import return prisma ready object

egemengunel committed:
refactor: copy function files from trigger/scraper/functions to cli/impoter/event

egemengunel committed:
refactor: update prisma routes, create index.ts

egemengunel committed:
refactor: rename scrape.ts to import_event.ts, remove trigger/scraper/functions route

egemengunel committed:
refactor: remove unused imports

egemengunel committed:
refactor: replace logger with console for debug statements,  add getOrg function to create profiles on schema_import, return prisma ready object,

egemengunel committed:
refactor: move getEventType,  getRedriectedUrl functions to index.ts

razbakov committed:
fix typo

egemengunel committed:
feat: add getEventType function export and create scrape_only test script

egemengunel committed:
refactor: move scrape testing to cli.ts file, sync schema.prisma with the global schema.prisma file. addtsk to package.json in /cli route

egemengunel committed:
refactor: move database event handling into index.ts

egemengunel committed:
refactor: -  rename getEventType  to fetchEvent

egemengunel committed:
move: move cloudinary, google_maps, linguist.ts files to cli/utils, rename import folder to import-event

egemengunel committed:
refactor: update getOrg function to return null instead of empty string, ensure photo property is defined, and streamline event data handling in facebook_import and schema_import files

egemengunel committed:
refactor: update CLI to use fetchEvent instead of getEventType for scraping event data

egemengunel committed:
fix: update utils/slug route

egemengunel committed:
feat: add event import and save command to CLI

egemengunel committed:
refacotr:  removeSourceUrl parameter from  importEvent function

egemengunel committed:
refactor: remove sourceUrl parameter from importEvent call in CLI

egemengunel committed:
refactor: move import-event outside of importer folder

razbakov commented on cli/import-event/facebook_import.ts:
```diff
@@ -0,0 +1,206 @@
+async function getOrg(
+  host: any,
+  place: any
+): Promise<null | {
```
this method should return either null or id of the organizer, not the object

razbakov commented on server/utils/google_maps.ts:
```diff
@@ -0,0 +1,88 @@
+import {
```
is this file used anywhere?

egemengunel replied:
i migrated this file from v3 because its funtions "getPlace" and "getCityId" are used in both the "facebook_import" and "schema_import" to get venue names and id's



 ``` let venue: any = ''

  if (venueName) {
    venue = await getPlace(venueName, venueCountry || 'de')
  }
  
  ```
  
  ```
  if (venue && !place) {
    place = await getCityId(venue)
  }
  ```
  
  
however when the import is completed, venueId always stays null even if location was specified in the event that was being imported

razbakov commented on trigger/import_event.ts:
```diff
@@ -0,0 +1,10 @@
+
+export const scrape = task({
+  id: 'import-event',
+  run: async (payload: any, { ctx }) => {
```
this task should explicitly show what it does:
- read event from database
- fetch event from url
- update event in database

it should be easy to read the code and see what happens
also this way you can use trigger logger

razbakov commented on pages/admin/events/index.vue:
```diff
@@ -103,7 +102,9 @@ const columns: ColumnDef<any>[] = [
     <div class="flex items-center gap-2">
       <Input v-model="searchQuery" placeholder="Search events..." />
-      <Button @click="handleCreateEvent"> Create Event </Button>
+      <NuxtLink to="/admin/events/import">
```
should be <Button as child>

razbakov commented on pages/admin/events/new.vue:
```diff
@@ -0,0 +1,27 @@
+    <Breadcrumb>
+      <BreadcrumbList>
+        <BreadcrumbItem>
+          <NuxtLink to="/admin/events"> Manage Events </NuxtLink>
```
should be:

<BreadcrumbLink as-child>
  <NuxtLink to="..."> ... </NuxtLink>
</BreadcrumbLink>

razbakov commented on pages/admin/events/import.vue:
```diff
@@ -0,0 +1,29 @@
+<script setup lang="ts">
+import Breadcrumb from '~/components/ui/breadcrumb/Breadcrumb.vue'
```
there is autoimport

razbakov commented on cli/utils/cloudinary.ts:
```diff
@@ -0,0 +1,22 @@
+  }
+
+  // For testing return original URL no cloudinary upload is made.
+  if (process.env.TEST_MODE === 'true') {
```
not a good way to do it

razbakov commented on cli/import-event/facebook_import.ts:
```diff
@@ -0,0 +1,206 @@
+import { scrapeFbEventFromFbid } from 'facebook-event-scraper'
```
missing dependency:

error TS2307: Cannot find module 'facebook-event-scraper' or its corresponding type declarations.

egemengunel committed:
refactor: simplify getOrg to return only the organizer ID and update related event handling in facebook_import

egemengunel committed:
refactor: update breadcrumb and button components for event management pages

egemengunel committed:
chore: add facebook-event-scraper dependency to CLI

razbakov commented on cli/cli.ts:
```diff
@@ -61,6 +63,45 @@ program.command('reindex').action(async (name) => {
 
+program.command('event:import:debug <url>').action(async (sourceUrl) => {
+  //test flag for cloudinary to return original url
+  process.env.TEST_MODE = 'true'
```
writing to process.env is not a good practice. `process.env` means it's environment configuration.

razbakov commented on cli/import-event/facebook_import.ts:
```diff
@@ -0,0 +1,189 @@
+
+  const orgFacebook = orgUrl.replace('https://www.facebook.com/', '')
+  if (orgFacebook) {
+    const profile = await prisma.profile.findFirst({
```
this variable is not used

egemengunel committed:
chore: clean up environment test flag for cloudinary

egemengunel committed:
refactor:remove unused Google Maps utility file left in /server/utils folder

egemengunel committed:
sync schema.prisma

egemengunel committed:
sync database with origin/main

egemengunel committed:
fix: remove manual import in index.ts

egemengunel committed:
refactor:
- move importEvent functionality to trigger
-remove handleImportError()
-remove saveEvent ()

egemengunel committed:
refactor: make name and slug fields optional,

egemengunel committed:
refactor: remove profile const from facebook_import

egemengunel committed:
refactor cli commands to adapt to moved importEvent functionality,

- add cli-tsx script to package json to be able to run the tests

-update schema in cli route

-remove unused variable 'failedImportSlug'

egemengunel committed:
migrate ticketPrice from v3

egemengunel committed:
fix: add helper for URL Constructor

egemengunel committed:
feat: add getVenue helper function to handle venue creation and retrieval

egemengunel committed:
feat: add getVenue function for venue profile creation and management in schema import

egemengunel committed:
chore: update dependencies in package.json and pnpm-lock.yaml, reintroducing @googlemaps/google-maps-services-js and adding cloudinary

egemengunel committed:
chore: move packages between correct locations

razbakov committed:
cleanup
